File: src/main/java/com/dongtronic/diabot/data/ConversionDTO.java
package com.dongtronic.diabot.data;

import com.dongtronic.diabot.converters.GlucoseUnit;
import org.jetbrains.annotations.NotNull;

public class ConversionDTO {
  private double original;
  private double mmol;
  private int mgdl;
  private GlucoseUnit inputUnit;


  /**
   * Create a ConversionDTO object with explicit unit
   *
   * @param original   original BG value
   * @param conversion converted BG value
   * @param inputUnit  original unit
   */
  public ConversionDTO(@NotNull double original, @NotNull double conversion, @NotNull GlucoseUnit inputUnit) {
    if (inputUnit == GlucoseUnit.AMBIGUOUS) {
      throw new IllegalArgumentException("single conversion constructor must contain explicit input unit");
    }

    setOriginal(original);
    this.inputUnit = inputUnit;

    if (inputUnit == GlucoseUnit.MMOL) {
      setMgdl(conversion);
      setMmol(original);
    } else if (inputUnit == GlucoseUnit.MGDL) {
      setMmol(conversion);
      setMgdl(original);
    }
  }

  /**
   * Create a ConversionDTO object with ambiguous unit
   *
   * @param original       original BG value
   * @param mmolConversion BG value converted to mmol/L
   * @param mgdlConversion BG value converted to mg/dL
   */
  public ConversionDTO(@NotNull double original, @NotNull double mmolConversion, @NotNull double mgdlConversion) {
    setOriginal(original);
    setMmol(mmolConversion);
    setMgdl(mgdlConversion);
    this.inputUnit = GlucoseUnit.AMBIGUOUS;
  }

  /**
   * Get the unit of the original BG value
   *
   * @return unit of the original BG value
   */
  public GlucoseUnit getInputUnit() {
    return inputUnit;
  }

  /**
   * Get the converted BG value
   *
   * @return double converted value
   * @throws IllegalStateException when the DTO contains an ambiguous conversion
   */
  public double getConverted() {
    if (inputUnit == GlucoseUnit.AMBIGUOUS) {
      throw new IllegalStateException("cannot retrieve specific unit result for ambiguous conversion");
    }

    if (inputUnit == GlucoseUnit.MGDL) {
      return mmol;
    } else {
      return mgdl;
    }
  }

  /**
   * Get the conversion result in mmol/L
   *
   * @return conversion result in mmol/L
   */
  public double getMmol() {
    return mmol;
  }

  /**
   * Get the conversion result in mg/dL
   *
   * @return conversion result in mg/dL
   */
  public int getMgdl() {
    return mgdl;
  }

  private void setMgdl(double input) {
    this.mgdl = (int) round(input, 0);
  }

  private void setMmol(double input) {
    this.mmol = round(input, 1);
  }

  public double getOriginal() {
    return original;
  }

  private void setOriginal(double original) {
    this.original = round(original, 1);
  }

  private static double round(double value, int precision) {
    int scale = (int) Math.pow(10, precision);
    return (double) Math.round(value * scale) / scale;
  }

  @Override
  public boolean equals(Object obj) {
    if (obj.getClass() != this.getClass()) {
      return false;
    }

    ConversionDTO other = (ConversionDTO) obj;

    if (other.getOriginal() != this.getOriginal()) {
      return false;
    }

    if (other.getConverted() != this.getConverted()) {
      return false;
    }

    if (other.getInputUnit() != this.getInputUnit()) {
      return false;
    }

    return true;

  }
}

File: utilities/src/main/java/com/dongtronic/diabot/data/ConversionDTO.kt
package com.dongtronic.diabot.data

import com.dongtronic.diabot.logic.diabetes.DiabetesConstants
import com.dongtronic.diabot.logic.diabetes.GlucoseUnit

/**
 * A data object for converted blood glucose values.
 */
class ConversionDTO {
    /**
     * The blood glucose that the conversion occurred from
     */
    val original: Number

    /**
     * The measurement unit that the conversion occurred from
     */
    val inputUnit: GlucoseUnit


    /**
     * The converted mmol/L blood glucose value
     */
    val mmol: Double

    /**
     * The converted mg/dL blood glucose value
     */
    val mgdl: Int

    /**
     * Create a ConversionDTO object with explicit unit
     *
     * @param original   original BG value
     * @param conversion converted BG value
     * @param inputUnit  original unit
     * @throws IllegalArgumentException If the [inputUnit] is [GlucoseUnit.AMBIGUOUS]
     */
    constructor(original: Number, conversion: Number, inputUnit: GlucoseUnit) {
        this.original = DiabetesConstants.round(original)
        this.inputUnit = inputUnit

        when (inputUnit) {
            GlucoseUnit.MMOL -> {
                this.mgdl = conversion.toInt()
                this.mmol = this.original.toDouble()
            }
            GlucoseUnit.MGDL -> {
                this.mgdl = this.original.toInt()
                this.mmol = DiabetesConstants.round(conversion.toDouble())
            }
            else -> throw IllegalArgumentException("Single conversion constructor must contain explicit input unit")
        }
    }

    /**
     * Create a ConversionDTO object with ambiguous measurement unit.
     *
     * @param original       original BG value
     * @param mmolConversion BG value converted to mmol/L
     * @param mgdlConversion BG value converted to mg/dL
     */
    constructor(original: Number, mmolConversion: Double, mgdlConversion: Int) {
        this.original = DiabetesConstants.round(original)
        this.inputUnit = GlucoseUnit.AMBIGUOUS
        this.mmol = DiabetesConstants.round(mmolConversion)
        this.mgdl = mgdlConversion
    }

    /**
     * Get the converted BG value. This essentially returns the opposite unit of the input unit.
     *
     * MG/DL  input -> MMOL/L output
     *
     * MMOL/L input -> MG/DL  output
     *
     * @return The converted BG value.
     *
     * If the input was mg/dl, this will be a mmol/L [Double] value.
     *
     * If the input was mmol/L, this will be a mg/dL [Int] value.
     * @throws IllegalStateException If the input unit is ambiguous.
     * */
    fun getConverted(): Number {
        return when (inputUnit) {
            GlucoseUnit.MGDL -> this.mmol
            GlucoseUnit.MMOL -> this.mgdl
            GlucoseUnit.AMBIGUOUS ->
                throw IllegalStateException("Cannot retrieve converted BG result for ambiguous input")
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ConversionDTO

        if (original != other.original) return false
        if (inputUnit != other.inputUnit) return false
        if (mmol != other.mmol) return false
        if (mgdl != other.mgdl) return false

        return true
    }

    override fun hashCode(): Int {
        var result = original.hashCode()
        result = 31 * result + inputUnit.hashCode()
        result = 31 * result + mmol.hashCode()
        result = 31 * result + mgdl
        return result
    }

    override fun toString(): String {
        return "ConversionDTO(original=$original, inputUnit=$inputUnit, mmol=$mmol, mgdl=$mgdl)"
    }
}